src/watcher.ts

```diff
@@ -28,245 +61,333 @@ import type {FailureMode} from './executor.js';
+      case 'running': {
+        this.#state = 'queued';
+        return;
+      }
+      case 'queued':
+      case 'aborted': {
+        return;
+      }
+      default: {
+        throw unknownState(this.#state);
+      }
+    }
+  };
 
-  /**
-   * Walk through a script config and return a list of absolute filesystem paths
-   * that we should watch for changes.
-   */
-  #getWatchPathGroups(
-    script: ScriptConfig
-  ): Array<{patterns: string[]; cwd: string}> {
-    const packageJsons = new Set<string>();
-    const groups: Array<{patterns: string[]; cwd: string}> = [];
+  #synchronizeInputFileWatchers(root: ScriptConfig): void {
     const visited = new Set<ScriptReferenceString>();
-
     const visit = (script: ScriptConfig) => {
       const key = scriptReferenceToString(script);
       if (visited.has(key)) {
         return;
       }
       visited.add(key);
-      packageJsons.add(pathlib.join(script.packageDir, 'package.json'));
-      if (script.files !== undefined) {
-        // TODO(aomarks) We could optimize to create fewer watchers, but we have
-        // to be careful to deal with "!"-prefixed negation entries, because
-        // those negations only apply to the previous entries **in that specific
-        // "files" array**. A simple solution could be that if a "files" array
-        // has any "!"-prefixed entry, then it gets its own watcher, otherwise
-        // we can group watchers by packageDir.
-        groups.push({patterns: script.files.values, cwd: script.packageDir});
+
+      const newInputFiles = script.files?.values;
+      const oldWatcher = this.#inputFileWatchers.get(key);
+      if (!watchPathsEqual(newInputFiles, oldWatcher?.patterns)) {
+        if (newInputFiles === undefined || newInputFiles.length === 0) {
+          this.#inputFileWatchers.delete(key);
+        } else {
+          const newWatcher = makeWatcher(
+            newInputFiles,
+            script.packageDir,
+            this.#fileChanged
+          );
+          this.#inputFileWatchers.set(key, newWatcher);
+        }
+        if (oldWatcher !== undefined) {
+          void oldWatcher.watcher.close();
+        }
       }
-      for (const dependency of script.dependencies) {
-        visit(dependency.config);
+      for (const dep of script.dependencies) {
+        visit(dep.config);
       }
     };
+    visit(root);
 
-    visit(script);
-    groups.push({
-      // The package.json group is already resolved to absolute paths, so cwd is
-      // arbitrary.
-      cwd: process.cwd(),
-      patterns: [...packageJsons],
-    });
-    return groups;
+    // There also could be some scripts that have been removed entirely.
+    for (const [oldKey, oldWatcher] of this.#inputFileWatchers) {
+      if (!visited.has(oldKey)) {
+        void oldWatcher.watcher.close();
+        this.#inputFileWatchers.delete(oldKey);
+      }
+    }
+  }
+
+  #onAbort(): void {
+    switch (this.#state) {
+      case 'watching':
+      case 'queued': {
```

Comment:
'queued' is like 'running', only we plan to start another run once it's finished yeah? In that case, I think we should treat it like 'running' rather than like 'watching' here.

as is, I think this could resolve #finished before the current execution is done